### PR TITLE
PMM-13533: Set up metadata v2 for new version fields.

### PR DESCRIPTION
**Pull request by idoqo**
Follow up from https://github.com/Percona-Lab/percona-version-service/pull/331;

This separates the metadata endpoint into `v1` and `v2`, allowing us to serve different responses on each of them. The underlying change is quite much, but at the API level, nothing should change (thus avoiding breaking changes for existing deployments).

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | failed |

File: server/metadata.go [hunk 1]
```diff
@@ -106,3 +148,22 @@ func (m *Metadata) parseFile(c []byte, fileExt string) (*pbVersion.MetadataVersi
...
+			return nil, errors.Join(err, errors.New("could not unmarshal yaml"))
+		}
+
+	case ".json":
+		if err := protojson.Unmarshal(c, meta); err != nil {
+			return nil, errors.Join(err, errors.New("could not unmarshal json"))
+		}
+	default:
+		return nil, fmt.Errorf("extension %s not supported", fileExt)
+	}
```
Comment (BupycHuk) [inline server/metadata.go:166]: this part is duplicated and I think we should enable allowing partial parsing and unknown fields in protojson

File: server/metadata.go [hunk 2]
```diff
@@ -26,66 +34,100 @@ func NewMetadata(fs fs.FS) (*Metadata, error) {
...
+			if err != nil {
+				return nil, nil, errors.Join(err, fmt.Errorf("could not read file %s", p))
+			}
+
+			metaV, err := m.parseV2File(c, filepath.Ext(f.Name()))
+			if err != nil {
+				return nil, nil, errors.Join(err, fmt.Errorf("could not parse file %s", f.Name()))
+			}
+			ret = append(ret, metaV)
 		}
```
Comment (BupycHuk) [inline server/metadata.go:112]: seems like we could use generics here to not duplicate code

File: server/metadata.go [hunk 3]
```diff
@@ -26,66 +34,100 @@ func NewMetadata(fs fs.FS) (*Metadata, error) {
...
+				m.v2Data[f.Name()] = &pbVersion.MetadataV2Response{Versions: v2Data}
+			}
+		}
 	}
 
 	return nil
 }
 
-func (m *Metadata) getAllMetadataFromFiles(product string) ([]*pbVersion.MetadataVersion, error) {
+func (m *Metadata) getAllMetadataFromFiles(product string, version metadataVersionDir) ([]*pbVersion.MetadataVersion, []*pbVersion.MetadataV2Version, error) {
```
Comment (BupycHuk) [inline server/metadata.go:87]: generics?

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Review by BupycHuk — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

File: server/metadata.go [hunk 4]
```diff
@@ -49,55 +60,68 @@ func (m *Metadata) readAll() error {
...
 		if err != nil {
-			return nil, errors.Join(err, fmt.Errorf("could not parse file %s", f.Name()))
+			return nil, nil, errors.Join(err, fmt.Errorf("could not parse file %s", f.Name()))
+		}
+		if metaV.ImageInfo == nil {
+			v1Data = append(v1Data, &pbVersion.MetadataVersion{
+				Version:     metaV.Version,
+				Recommended: metaV.Recommended,
+				Supported:   metaV.Supported,
+			})
```
Comment (BupycHuk) [inline server/metadata.go:104]: I think we need to add it to v2Data as well

Comment (idoqo) [reply server/metadata.go:104]: yeah that makes sense.

Comment (idoqo) [reply server/metadata.go:104]: Yeah that makes sense

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

File: server/metadata.go [hunk 5]
```diff
@@ -49,55 +60,64 @@ func (m *Metadata) readAll() error {
...
+			return nil, nil, errors.Join(err, fmt.Errorf("could not read file %s", p))
 		}
 
 		metaV, err := m.parseFile(c, filepath.Ext(f.Name()))
 		if err != nil {
-			return nil, errors.Join(err, fmt.Errorf("could not parse file %s", f.Name()))
+			return nil, nil, errors.Join(err, fmt.Errorf("could not parse file %s", f.Name()))
+		}
+		v2Data = append(v2Data, metaV)
+		if metaV.ImageInfo == nil {
```
Comment (BupycHuk) [inline server/metadata.go:97]: I think we can drop this check

**Review by BupycHuk — APPROVED** [review]

**Review by recharte — DISMISSED** [review]
Looks good :slightly_smiling_face:

File: server/metadata.go [hunk 6]
```diff
@@ -49,55 +60,64 @@ func (m *Metadata) readAll() error {
...
+			v1Data = append(v1Data, &pbVersion.MetadataVersion{
+				Version:     metaV.Version,
+				Recommended: metaV.Recommended,
+				Supported:   metaV.Supported,
+			})
 		}
```
Comment (BupycHuk) [inline server/metadata.go:103]:
```suggestion
		v1Data = append(v1Data, &pbVersion.MetadataVersion{
			Version:     metaV.Version,
			Recommended: metaV.Recommended,
			Supported:   metaV.Supported,
		})
```

**Review by BupycHuk — APPROVED** [review]

**Review by ademidoff — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

File: server/metadata.go (hunk 5 above)
Comment (idoqo) [reply server/metadata.go]: Why though? 🤔

**Review by BupycHuk — DISMISSED** [review]

File: api/version.proto [hunk 7]
```diff
@@ -67,6 +67,17 @@ service VersionService {
...
   }
 
+  // Metadata v2 provides metadata information about products. It is an extension of Metadata with new fields.
+  rpc MetadataV2(MetadataRequest) returns (MetadataV2Response) {
+    option (google.api.http) = {
+      get: "/metadata/v2/{product}"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "v2 metadata for a product"
+      description: "Return metadata information with additional image information for a product"
```
Comment (BupycHuk) [inline api/version.proto:77]:
```suggestion
      description: "Return metadata information based with additional image information (v2) for a product"
```

**Review by ademidoff — DISMISSED** [review]

File: api/version.proto [hunk 8]
```diff
@@ -67,6 +67,17 @@ service VersionService {
...
   }
 
+  // Metadata v2 provides metadata information about products. It is an extension of Metadata with new fields.
+  rpc MetadataV2(MetadataRequest) returns (MetadataV2Response) {
+    option (google.api.http) = {
+      get: "/metadata/v2/{product}"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "v2 metadata for a product"
+      description: "Return metadata information based with additional image information (v2) for a product"
```
Comment (BupycHuk) [inline api/version.proto:77]:
```suggestion
      description: "Return metadata information with additional image information for a product"
```

File: api/version.swagger.yaml [hunk 9]
```diff
@@ -33,6 +33,27 @@ paths:
           type: string
       tags:
         - VersionService
+  /metadata/v2/{product}:
+    get:
+      summary: v2 metadata for a product
+      description: Return metadata information based with additional image information (v2) for a product
```
Comment (BupycHuk) [inline api/version.swagger.yaml:39]:
```suggestion
      description: Return metadata information with additional image information for a product
```

File: third_party/OpenAPI/api/version.swagger.yaml [hunk 10]
```diff
@@ -33,6 +33,27 @@ paths:
           type: string
       tags:
         - VersionService
+  /metadata/v2/{product}:
+    get:
+      summary: v2 metadata for a product
+      description: Return metadata information based with additional image information (v2) for a product
```
Comment (BupycHuk) [inline third_party/OpenAPI/api/version.swagger.yaml:39]:
```suggestion
      description: Return metadata information with additional image information for a product
```

**Review by ademidoff — APPROVED** [review]

**Review by BupycHuk — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Pull request merged**